cdap-formats/src/main/java/io/cdap/cdap/format/io/JsonStructuredRecordDatumWriter.java

```diff
@@ -132,61 +144,89 @@ protected void encode(Encoder encoder, Schema schema, Object value) throws IOExc
-          encoder.writeString(value.toString());
-          break;
+          encodeDecimal(encoder, bigDecimal);
+          return;
+        }
       }
-      return;
     }
     super.encode(encoder, schema, value);
   }
 
+  protected void handleLogicalTypeAsString(Encoder encoder,
+                                           Object value,
+                                           Schema nonNullableSchema,
+                                           Schema.LogicalType logicalType) throws IOException {
+    switch (logicalType) {
+      case DATE:
+        Integer date = (Integer) value;
+        // will be encoded to string of format YYYY-mm-DD
+        encoder.writeString(LocalDate.ofEpochDay(date.longValue()).format(DateTimeFormatter.ISO_LOCAL_DATE));
+        break;
+      case TIME_MILLIS:
+        LocalTime localTimeMillis = LocalTime.ofNanoOfDay(TimeUnit.MILLISECONDS.toNanos(((Integer) value)));
+        // will be encoded to string of format HH:mm:ss.SSSSSSSSS
+        encoder.writeString(localTimeMillis.format(DateTimeFormatter.ISO_LOCAL_TIME));
+        break;
+      case TIME_MICROS:
+        LocalTime localTimeMicros = LocalTime.ofNanoOfDay(TimeUnit.MICROSECONDS.toNanos((Long) value));
+        // will be encoded to string of format HH:mm:ss.SSSSSSSSS
+        encoder.writeString(localTimeMicros.format(DateTimeFormatter.ISO_LOCAL_TIME));
+        break;
+      case TIMESTAMP_MILLIS:
+        ZonedDateTime timestampMillis = getZonedDateTime((Long) value, TimeUnit.MILLISECONDS,
+                                                         ZoneId.ofOffset("UTC", ZoneOffset.UTC));
+        // will be encoded to string of format YYYY-mm-DDTHH:mm:ss.SSSSSSSSSZ[UTC]
+        encoder.writeString(timestampMillis.format(DateTimeFormatter.ISO_ZONED_DATE_TIME));
+        break;
+      case TIMESTAMP_MICROS:
+        ZonedDateTime timestampMicros = getZonedDateTime((Long) value, TimeUnit.MICROSECONDS,
+                                                         ZoneId.ofOffset("UTC", ZoneOffset.UTC));
+        // will be encoded to string of format YYYY-mm-DDTHH:mm:ss.SSSSSSSSSZ[UTC]
+        encoder.writeString(timestampMicros.format(DateTimeFormatter.ISO_ZONED_DATE_TIME));
+        break;
+      case DECIMAL:
+        int scale = nonNullableSchema.getScale();
+        BigDecimal bigDecimal;
+        if (value instanceof ByteBuffer) {
+          bigDecimal = new BigDecimal(new BigInteger(Bytes.toBytes((ByteBuffer) value)), scale);
+        } else {
+          bigDecimal = new BigDecimal(new BigInteger((byte[]) value), scale);
+        }
+        encoder.writeString(bigDecimal.toString());
+        break;
+      case DATETIME:
+        //expecting value to be already in the correct format
+        encoder.writeString(value.toString());
+        break;
+    }
+  }
+
+  @Override
+  protected void encodeBytes(Encoder encoder, Object value) throws IOException {
+    encoder.writeString(FormatUtils.base64Encode(value));
```

Comment:
It's better to do in json encoder. Do we have approval from product to do this breaking change? Current approach is perfectly valid, so a lot of customers may use it already

Reply:
Removed this piece of the PR